File: cmd/verify.ts
#!/usr/bin/env bun
// verify.ts -- verify djvu_test render output against DjVuLibre's ddjvu, page
// by page (run with `bun cmd/verify.ts`; replaces the old Python verify.py).
//
// This is the test entry point: it ensures deps (get-deps.ts), builds the ref
// tools + harness (buildRef/build from build.ts), then verifies.
//
// Pages that are pure JB2 masks (Sjbz, no BG44/FG44 background) must match
// `ddjvu -format=pgm` byte-for-byte. Pages with an IW44 background or color are
// compared as ppm. Text is compared against djvutxt (trailing separators
// ignored). Runs over every .djvu under testfiles/ recursively; set the
// DJVU_SPECS env var to point the scan at a different directory.
import { existsSync, readFileSync, readdirSync, rmSync, statSync } from "fs";
import { tmpdir } from "os";
import { join, dirname, basename } from "path";
import { getDeps } from "./get-deps";
import { buildRef, build, defaultUseClang } from "./build";

const ROOT = dirname(import.meta.dir);
const RB = join(ROOT, "ref_build");
let TEST = join(ROOT, "djvu_test.exe"); // set to the built exe in main()
const TMP = tmpdir();

const tag = (d: Buffer, p: number) => d.toString("latin1", p, p + 4);

// File offsets of every displayable page FORM (FORM:DJVU).
function pageOffsets(data: Buffer): number[] {
  let p = tag(data, 0) === "AT&T" ? 4 : 0;
  if (tag(data, p) !== "FORM") return [];
  if (tag(data, p + 8) === "DJVU") return [p];
  const end = p + 8 + data.readUInt32BE(p + 4);
  let q = p + 12;
  while (q + 8 <= end) {
    const cid = tag(data, q);
    const csz = data.readUInt32BE(q + 4);
    if (cid === "DIRM") {
      const d = q + 8;
      const cnt = data.readUInt16BE(d + 1);
      const offs: number[] = [];
      for (let i = 0; i < cnt; i++) offs.push(data.readUInt32BE(d + 3 + i * 4));
      return offs.filter(
        (o) => tag(data, o) === "FORM" && tag(data, o + 8) === "DJVU",
      );
    }
    q += 8 + csz + (csz & 1);
  }
  return [];
}

// 'mask' (pure Sjbz), 'bg' (has an IW44/JPEG background), or 'other'.
function pageKind(data: Buffer, off: number): "mask" | "bg" | "other" {
  const end = off + 8 + data.readUInt32BE(off + 4);
  let p = off + 12;
  let bg = false,
    sj = false;
  while (p + 8 <= end) {
    const cid = tag(data, p);
    const csz = data.readUInt32BE(p + 4);
    if (cid === "BG44" || cid === "FG44" || cid === "BGjp" || cid === "FGjp") bg = true;
    if (cid === "Sjbz") sj = true;
    p += 8 + csz + (csz & 1);
  }
  if (bg) return "bg";
  if (sj) return "mask";
  return "other";
}

// Whether any page FORM carries annotation (ANTa/ANTz) or text (TXTa/TXTz)
// chunks -- a quick hint for the progress line.
function docFeatures(data: Buffer, offs: number[]): { anno: boolean; text: boolean } {
  let anno = false,
    text = false;
  for (const off of offs) {
    const end = off + 8 + data.readUInt32BE(off + 4);
    let p = off + 12;
    while (p + 8 <= end) {
      const cid = tag(data, p);
      const csz = data.readUInt32BE(p + 4);
      if (cid === "ANTa" || cid === "ANTz") anno = true;
      if (cid === "TXTa" || cid === "TXTz") text = true;
      p += 8 + csz + (csz & 1);
    }
  }
  return { anno, text };
}

function run(cmd: string[]): Buffer {
  const r = Bun.spawnSync({ cmd, stdout: "pipe", stderr: "pipe" });
  return Buffer.from(r.stdout);
}

function renderRef(f: string, page: number, out: string, fmt = "pgm") {
  run([join(RB, "ddjvu.exe"), `-format=${fmt}`, `-page=${page}`, f, out]);
}
function renderMine(f: string, page: number, out: string) {
  if (existsSync(out)) rmSync(out);
  run([TEST, "-page", String(page), "-out", out, f]);
}

// strip CR / form-feed and trailing whitespace (text normalization)
function textNorm(b: Buffer): string {
  return b
    .toString("latin1")
    .replace(/\r/g, "")
    .replace(/\f/g, "")
    .replace(/\s+$/, "");
}
const getTextRef = (f: string, page: number) =>
  run([join(RB, "djvutxt.exe"), `--page=${page}`, f]);
const getTextMine = (f: string, page: number) =>
  run([TEST, "-page", String(page), "-text", f]);

function readBytes(p: string): Buffer {
  return existsSync(p) ? readFileSync(p) : Buffer.alloc(0);
}

// Every .djvu under dir, recursively (sorted by path).
function walkDjvu(dir: string): string[] {
  if (!existsSync(dir)) return [];
  const out: string[] = [];
  for (const name of readdirSync(dir)) {
    const p = join(dir, name);
    if (statSync(p).isDirectory()) out.push(...walkDjvu(p));
    else if (name.toLowerCase().endsWith(".djvu")) out.push(p);
  }
  return out;
}

async function main(): Promise<number> {
  // Ensure the corpus + reference checkouts exist, then build everything.
  await getDeps();
  // Verify against every .djvu under testfiles/ (recursively); DJVU_SPECS
  // overrides the root to point at any other directory of samples.
  const SPECS = process.env.DJVU_SPECS ?? join(ROOT, "testfiles");
  // -clang forces the clang harness; default is MSVC on Windows.
  const useClang = process.argv.includes("-clang") || defaultUseClang;
  await buildRef();
  TEST = await build(useClang);

  let m = 0,
    mm = 0,
    skip = 0;
  let tm = 0,
    tmm = 0,
    te = 0;
  const bad: string[] = [];
  const tbad: string[] = [];

  const files = walkDjvu(SPECS).sort();

  const ref = join(TMP, "djref.pnm");
  const mine = join(TMP, "djmine.pnm");

  files.forEach((f, fi) => {
    const data = readFileSync(f);
    const offs = pageOffsets(data);
    const { anno, text } = docFeatures(data, offs);
    const feats = [`${offs.length} pages`];
    if (anno) feats.push("annots");
    if (text) feats.push("text");
    console.log(`[${fi + 1}/${files.length}] ${basename(f)} (${feats.join(", ")})`);
    offs.forEach((o, i) => {
      const page = i + 1;
      // render: pure-mask pages -> pgm (gray); bg/color pages -> ppm
      const kind = pageKind(data, o);
      if (kind === "other") {
        skip++;
      } else {
        const fmt = kind === "mask" ? "pgm" : "ppm";
        renderRef(f, page, ref, fmt);
        renderMine(f, page, mine);
        const a = readBytes(ref);
        const b = readBytes(mine);
        if (a.length && a.equals(b)) {
          m++;
        } else {
          mm++;
          bad.push(`${basename(f)} p${page} ${kind} (ref=${a.length} mine=${b.length})`);
        }
      }
      // text (all pages; ignores trailing page separator)
      const rt = textNorm(getTextRef(f, page));
      const mt = textNorm(getTextMine(f, page));
      if (!rt && !mt) te++;
      else if (rt === mt) tm++;
      else {
        tmm++;
        tbad.push(`${basename(f)} p${page}`);
      }
    });
  });

  console.log(`render (mask=pgm, bg/color=ppm): MATCH=${m} MISMATCH=${mm}; skipped=${skip}`);
  for (const x of bad.slice(0, 50)) console.log("  render MISMATCH", x);
  console.log(`text: MATCH=${tm} MISMATCH=${tmm}; both-empty=${te}`);
  for (const x of tbad.slice(0, 50)) console.log("  text MISMATCH", x);
  return mm || tmm ? 1 : 0;
}

process.exit(await main());

File: test/verify.py
#!/usr/bin/env python3
"""Verify djvu_test render output against DjVuLibre's ddjvu, page by page.

Pages that are pure JB2 masks (Sjbz, no BG44/FG44 background) must match
ddjvu -format=pgm byte-for-byte. Pages with an IW44 background or color are
reported separately (those need the IW44 codec, not yet implemented)."""
import struct, subprocess, os, glob, sys, tempfile

SPECS = os.environ.get("DJVU_SPECS", r"C:\Users\kjk\src\DjvuNet\Specs")
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
RB = os.path.join(ROOT, "ref_build")
TEST = os.path.join(ROOT, "djvu_test.exe")
TMP = tempfile.gettempdir()

def page_offsets(data):
    p = 4 if data[:4] == b'AT&T' else 0
    if data[p:p+4] != b'FORM':
        return []
    ftype = data[p+8:p+12]
    if ftype == b'DJVU':
        return [p]
    end = p + 8 + struct.unpack('>I', data[p+4:p+8])[0]
    q = p + 12
    while q + 8 <= end:
        cid = data[q:q+4]; csz = struct.unpack('>I', data[q+4:q+8])[0]
        if cid == b'DIRM':
            d = q + 8; cnt = struct.unpack('>H', data[d+1:d+3])[0]
            offs = [struct.unpack('>I', data[d+3+i*4:d+7+i*4])[0] for i in range(cnt)]
            return [o for o in offs
                    if data[o:o+4] == b'FORM' and data[o+8:o+12] == b'DJVU']
        q += 8 + csz + (csz & 1)
    return []

def page_kind(data, off):
    """returns 'mask' (pure Sjbz), 'bg' (has IW44 background), or 'other'"""
    size = struct.unpack('>I', data[off+4:off+8])[0]
    end = off + 8 + size; p = off + 12
    bg = sj = False
    while p + 8 <= end:
        cid = data[p:p+4]; csz = struct.unpack('>I', data[p+4:p+8])[0]
        if cid in (b'BG44', b'FG44', b'BGjp', b'FGjp'): bg = True
        if cid == b'Sjbz': sj = True
        p += 8 + csz + (csz & 1)
    if bg: return 'bg'
    if sj: return 'mask'
    return 'other'

def render_ref(f, page, out, fmt="pgm"):
    subprocess.run([os.path.join(RB, "ddjvu.exe"), f"-format={fmt}",
                    f"-page={page}", f, out], capture_output=True)

def render_mine(f, page, out):
    if os.path.exists(out): os.remove(out)
    subprocess.run([TEST, "-page", str(page), "-out", out, f], capture_output=True)

def text_norm(b):
    return b.replace(b'\r', b'').replace(b'\x0c', b'').rstrip()

def get_text_ref(f, page):
    return subprocess.run([os.path.join(RB, "djvutxt.exe"), f"--page={page}", f],
                          capture_output=True).stdout

def get_text_mine(f, page):
    return subprocess.run([TEST, "-page", str(page), "-text", f],
                          capture_output=True).stdout

def main():
    m = mm = skip = 0
    tm = tmm = te = 0
    bad = []; tbad = []
    for f in sorted(glob.glob(os.path.join(SPECS, "*.djvu"))):
        data = open(f, "rb").read()
        for i, o in enumerate(page_offsets(data)):
            # render: pure-mask pages -> pgm (gray); bg/color pages -> ppm
            kind = page_kind(data, o)
            if kind == 'other':
                skip += 1
            else:
                fmt = "pgm" if kind == 'mask' else "ppm"
                ref = os.path.join(TMP, "djref.pnm"); mine = os.path.join(TMP, "djmine.pnm")
                render_ref(f, i+1, ref, fmt)
                render_mine(f, i+1, mine)
                a = open(ref, "rb").read() if os.path.exists(ref) else b""
                b = open(mine, "rb").read() if os.path.exists(mine) else b""
                if a and a == b:
                    m += 1
                else:
                    mm += 1
                    bad.append(f"{os.path.basename(f)} p{i+1} {kind} (ref={len(a)} mine={len(b)})")
            # text (all pages; ignores trailing page separator)
            rt = text_norm(get_text_ref(f, i+1))
            mt = text_norm(get_text_mine(f, i+1))
            if not rt and not mt:
                te += 1
            elif rt == mt:
                tm += 1
            else:
                tmm += 1
                tbad.append(f"{os.path.basename(f)} p{i+1}")
    print(f"render (mask=pgm, bg/color=ppm): MATCH={m} MISMATCH={mm}; skipped={skip}")
    for x in bad[:50]:
        print("  render MISMATCH", x)
    print(f"text: MATCH={tm} MISMATCH={tmm}; both-empty={te}")
    for x in tbad[:50]:
        print("  text MISMATCH", x)
    return 1 if (mm or tmm) else 0

if __name__ == "__main__":
    sys.exit(main())
